caffe2/operators/onnxifi_op.cc

```diff
@@ -151,42 +151,63 @@ bool OnnxifiOp<float, CPUContext>::RunOnDevice() {
     SetOutputTensorDescriptorTypeAndBuffer(
         type, output_tensor, &tensor_descriptor);
   }
-
-  CAFFE_ENFORCE_EQ(
-      lib_->onnxSetGraphIO(
-          graph_,
-          input_desc_.size(),
-          input_desc_.data(),
-          output_desc_.size(),
-          output_desc_.data()),
-      ONNXIFI_STATUS_SUCCESS);
-
-  onnxMemoryFenceV1 input_fence;
-  input_fence.tag = ONNXIFI_TAG_MEMORY_FENCE_V1;
-  input_fence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
-  CAFFE_ENFORCE_EQ(
-      lib_->onnxInitEvent(backend_, &input_fence.event),
-      ONNXIFI_STATUS_SUCCESS);
-  onnxMemoryFenceV1 output_fence;
-  output_fence.tag = ONNXIFI_TAG_MEMORY_FENCE_V1;
-  output_fence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
-
-  // Call the asycn run on backend, singal event on input fence and wait for the
-  // event on output fence
-  CAFFE_ENFORCE_EQ(
-      lib_->onnxSignalEvent(input_fence.event), ONNXIFI_STATUS_SUCCESS);
-  CAFFE_ENFORCE_EQ(
-      lib_->onnxRunGraph(graph_, &input_fence, &output_fence),
-      ONNXIFI_STATUS_SUCCESS);
-  CAFFE_ENFORCE_EQ(
-      lib_->onnxWaitEvent(output_fence.event), ONNXIFI_STATUS_SUCCESS);
-
-  // Destroy the event objects
-  CAFFE_ENFORCE_EQ(
-      lib_->onnxReleaseEvent(input_fence.event), ONNXIFI_STATUS_SUCCESS);
-  CAFFE_ENFORCE_EQ(
-      lib_->onnxReleaseEvent(output_fence.event), ONNXIFI_STATUS_SUCCESS);
-
+  bool ext_success = false;
+#ifdef ONNXIFI_ENABLE_EXT
+  onnxExtensionFunctionPointer onnxSetIOAndRunGraphPointer;
+  if (lib_->onnxGetExtensionFunctionAddress(
+          backend_id_, "onnxSetIOAndRunGraph", onnxSetOpAndRunGraphPointer) ==
+      ONNXIFI_STATUS_SUCCESS) {
+    onnxMemoryFenceV1 output_fence;
+    output_fence.tag = ONNXIFI_TAG_MEMORY_FENCE_V1;
+    output_fence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
+    CAFFE_ENFORCE_EQ(
+        (*onnxExtensionFunctionPointer)(
+            graph,
+            input_desc_.size(),
+            input_desc_.data(),
+            output_desc_.size(),
+            output_desc_.data(),
+            output_fence),
```

Comment:
you still have to wait on this, release the event, etc

Reply:
I thought this is to be done in SetIOAndRunGraph. Okay then.

Reply:
The output event signals when the data is ready to be read by caffe2, Glow will signal the event when it's good to go but if caffe2 doesn't wait for the event to be signalled it will jump the gun and try to read output before it's ready.

Reply:
@jackm321 So I guess I only need to deal with output fence event but not input fence event, right? Since there is no input fence here.
Thanks!